### PR TITLE
Implemented performance improvement for TriviaHelper.GetContainingTriviaList 

### LINKED ISSUE #957: Improve performance of TriviaHelper.GetContainingTriviaList
TriviaHelper.GetContainingTriviaList is currently responsible for 15.4% of the inclusive samples in my cpu sampling. I think this is way to much. The most expensive call in it is actually TriviaList.AddRange with 13.3%. The callers of this method don't really need a TriviaList so it might be a good idea to switch to a different list.

### PR DESCRIPTION
**Pull request by vweijsters**
Fixes #957 

Tested it using the StyleCopTester pull request.
In release mode, running SA1512 over the Microsoft/CodeContracts solution resulted in:

Scenario | Result (time spent in TriviaHelper.GetContainingTriviaList)
----- | -----
Original | 12764 ms (average)
New | 1962 ms (average)

Big thanks to @pdelvo for helping out (= doing a lot of the work) with this one.

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs [hunk 1]
```diff
@@ -14,20 +17,20 @@ internal static class TriviaHelper
...
         {
             for (var index = 0; index < triviaList.Count; index++)
             {
                 var currentTrivia = triviaList[index];
                 switch (currentTrivia.Kind())
                 {
-                case SyntaxKind.EndOfLineTrivia:
-                case SyntaxKind.WhitespaceTrivia:
-                    break;
+                    case SyntaxKind.EndOfLineTrivia:
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs]: You should probably revert thise indentation changes

File: StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1512SingleLineCommentsMustNotBeFollowedByBlankLine.cs [hunk 2]
```diff
@@ -170,9 +171,9 @@ private static bool IsOnOwnLine(SyntaxTriviaList triviaList, int triviaIndex)
             return false;
         }
 
-        private static bool IsPartOfFileHeader(SyntaxTriviaList triviaList, int triviaIndex)
+        private static bool IsPartOfFileHeader(IReadOnlyList<SyntaxTrivia> triviaList, int triviaIndex)
         {
-            if (triviaList.FullSpan.Start > 0)
+            if (triviaList[0].FullSpan.Start > 0)
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1512SingleLineCommentsMustNotBeFollowedByBlankLine.cs]: Can you be sure that triviaList is not empty?

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs (hunk 1 above)
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs]: Will do. The auto formatter of Visual Studio got me again :disappointed:

File: StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1512SingleLineCommentsMustNotBeFollowedByBlankLine.cs (hunk 2 above)
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1512SingleLineCommentsMustNotBeFollowedByBlankLine.cs]: No, probably not.

**Comment by pdelvo** [comment]
Thats a really significant improvement. Using the StyleCopTester with roslyn 1.0.0 running all our analyzers on Microsoft/CodeContracts gave me these results (green arrows are improvements)

comparison

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs [hunk 3]
```diff
@@ -130,19 +133,30 @@ internal static SyntaxTriviaList WithoutLeadingWhitespace(this SyntaxTriviaList
...
                 var nextToken = token.GetNextToken(includeZeroWidth: true);
-                return token.TrailingTrivia.AddRange(nextToken.LeadingTrivia);
+
+                part1 = token.TrailingTrivia;
+                part2 = nextToken.LeadingTrivia;
+            }
+            else
+            {
+                var prevToken = token.GetPreviousToken();
+                triviaIndex = prevToken.TrailingTrivia.Count + token.LeadingTrivia.IndexOf(trivia);
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs]: triviaIndex is not used. This is actually a (relativly) big performance hit. Thats how I noticed it.

Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs]: :question: Is this something that still needs to be fixed before the change is merged?

Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs]: You could remove this when merging. This shows up prominently in performance reports

Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs]: :exclamation: You cannot remote that. `triviaIndex` is an out parameter and several callers of the method require that value.

**Comment by vweijsters** [comment]
I also removed all instances of SyntaxTriviaList.AddRange. 
This will address some of the issues in #964.

I added it in the PR because it needed the helper class introduced here.

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs (hunk 3 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs]: You're right. I overlooked that

**Pull request merged**